src/watcherEx.ts

```diff
@@ -0,0 +1,227 @@
+   * @example
+   * newWatcherEx('redis://user:password@redis-service.com:6379')
+   * newWatcherEx('//localhost:6379')
+   * newWatcherEx({
+   *   port: 6379, // Redis port
+   *   host: "127.0.0.1", // Redis host
+   *   family: 4, // 4 (IPv4) or 6 (IPv6)
+   *   password: "auth",
+   *   db: 0,
+   *   channel: "casbin"
+   * })
+   */
+  public static async newWatcherEx(
+    options?: WatcherExOptions | string,
+    callBackOptions?: WatcherExCallBackOptions
+  ): Promise<RedisWatcherEx> {
+    let channel = '';
+    if (typeof options === 'object' && options.channel) {
+      channel = options.channel;
+    }
+
+    const pubConnection = new RedisConnection(options);
+    const subConnection = new RedisConnection(options);
+    return this.init(pubConnection, subConnection, channel, callBackOptions);
+  }
+
+  /**
+   * newWatcherExWithCluster creates a watcherEx on the Redis cluster.
+   * @param nodes - An array of nodes in the cluster, [{ port: number, host: string }]
+   * @param clusterOptions - https://github.com/luin/ioredis/blob/master/API.md#new-clusterstartupnodes-options
+   * @example
+   * newWatcherExWithCluster([{ port: 6380, host: "127.0.0.1"}, { port: 6381, host: "127.0.0.1"})
+   */
+  public static async newWatcherExWithCluster(
+    nodes: ClusterNode[] = [],
+    clusterOptions: WatcherExClusterOptions = {},
+    callBackOptions?: WatcherExCallBackOptions
+  ): Promise<RedisWatcherEx> {
+    const pubConnection = new RedisClusterConnection(nodes, clusterOptions);
+    const subConnection = new RedisClusterConnection(nodes, clusterOptions);
+    return this.init(pubConnection, subConnection, clusterOptions.channel, callBackOptions);
+  }
+
+  private static async init(
+    pubConnection: RedisConnection | RedisClusterConnection,
+    subConnection: RedisConnection | RedisClusterConnection,
+    channel?: string,
+    callBackOptions?: WatcherExCallBackOptions
+  ): Promise<RedisWatcherEx> {
+    const watcherEx = new RedisWatcherEx();
+
+    watcherEx.pubConnection = pubConnection;
+    watcherEx.subConnection = subConnection;
+
+    if (channel) {
+      watcherEx.channel = channel;
+    }
+
+    watcherEx.watcherExCallBackOptions = callBackOptions;
+    const watcherExCallBackMethods: any = {
+      updateForAddPolicy: callBackOptions?.addPolicyCb,
+      updateForRemovePolicy: callBackOptions?.removePolicyCb,
+      updateForRemoveFilteredPolicy: callBackOptions?.removeFilteredPolicyCb,
+      updateForSavePolicy: callBackOptions?.savePolicyCb,
+      updateForAddPolicies: callBackOptions?.addPoliciesCb,
+      updateForRemovePolicies: callBackOptions?.removePoliciesCb
+    };
+
+    const client = await watcherEx.subConnection.getRedisClient();
+    // @ts-ignore - subscribe is exists.
+    await client.subscribe(watcherEx.channel);
+    client.on('message', (chan: string, msg: any) => {
+      if (chan !== watcherEx.channel) {
+        return;
+      }
+      msg = JSON.parse(msg);
+      try {
```

Comment:
Maybe do not throw errors unnecessary. 

`if(watcherExCallBackMethods[msg.updateType]) watcherExCallBackMethods[msg.updateType](msg.args);` or `watcherExCallBackMethods[msg.updateType]?.call(this, msg.args);`

Reply:
but it is not throwing error in try& catch block I have added test for missing callback

Reply:
Empty `catch` is not really a good way to use these functions. In addition, if something failed within the callback function itself, it would not be caught anymore.